cdap-app-templates/cdap-etl/cdap-etl-realtime/src/main/java/co/cask/cdap/templates/etl/realtime/ETLWorker.java

```diff
@@ -172,47 +174,82 @@ private void initializeSink(WorkerContext context, ETLStage stage) throws Except
 
   @Override
   public void run() {
+    final SourceState currentState = new SourceState();
+    final SourceState nextState = new SourceState();
+    final List<Object> dataToSink = Lists.newArrayList();
     running = true;
+
+    // Fetch SourceState from State Table.
+    // Only required at the beginning since we persist the state if there is a change.
+    getContext().execute(new TxRunnable() {
+      @Override
+      public void run(DatasetContext context) throws Exception {
+        KeyValueTable stateTable = context.getDataset(ETLRealtimeTemplate.STATE_TABLE);
+        byte[] stateBytes = stateTable.read(stateStoreKeyBytes);
+        if (stateBytes != null) {
+          SourceState state = GSON.fromJson(Bytes.toString(stateBytes), SourceState.class);
+          currentState.setState(state);
+        }
+      }
+    });
+
     while (running) {
-      final SourceState sourceState = new SourceState();
-
-      // Fetch SourceState from State Table.
-      getContext().execute(new TxRunnable() {
-        @Override
-        public void run(DatasetContext context) throws Exception {
-          KeyValueTable stateTable = context.getDataset(ETLRealtimeTemplate.STATE_TABLE);
-          byte[] stateBytes = stateTable.read(stateStoreKeyBytes);
-          if (stateBytes != null) {
-            SourceState state = GSON.fromJson(Bytes.toString(stateBytes), SourceState.class);
-            sourceState.setState(state.getState());
-          }
+      // Invoke poll method of the source to fetch data
+      try {
+        SourceState newState = source.poll(sourceEmitter, new SourceState(currentState));
+        if (newState != null) {
+          nextState.setState(newState);
         }
-      });
+      } catch (Exception e) {
+        // Continue since the source threw an exception. No point in processing records and state is not changed.
+        LOG.warn("Adapter {} : Exception thrown during polling of Source for data", adapterName, e);
+        sourceEmitter.reset();
+        continue;
+      }
 
-      final SourceState nextState = source.poll(sourceEmitter, sourceState);
+      // For each object emitted by the source, invoke the transformExecutor and collect all the data
+      // to be persisted in the sink.
       for (Object sourceData : sourceEmitter) {
         try {
-          final Iterable<Object> dataToSink = transformExecutor.runOneIteration(sourceData);
+          for (Object object : transformExecutor.runOneIteration(sourceData)) {
+            dataToSink.add(object);
+          }
+        } catch (Exception e) {
+          LOG.warn("Adapter {} : Exception thrown while processing data {}", adapterName, sourceData, e);
+        }
+      }
+      sourceEmitter.reset();
 
+      // Start a Transaction if there is data to persist or if the Source state has changed.
+      try {
+        if ((dataToSink.size() != 0) || (!nextState.equals(currentState))) {
```

Comment:
`dataToSink.size()` can be replaced by `dataToSink.empty()` here and and in line 231.